File: light-fantastic/src/container.ts
import { Container as CommonContainer } from "powerpi-common";
import { Constructable, Container as LocalContainer } from "typedi";

export default class Container {
    public static get<TService>(identifier: Constructable<TService>): TService {
        try {
            return CommonContainer.get(identifier);
        } catch {
            return LocalContainer.get(identifier);
        }
    }

    public static set<TService>(identifier: Constructable<TService>, service: TService) {
        CommonContainer.set(identifier, service);
    }

    public static override<TParent, TOverride extends TParent>(
        parent: Constructable<TParent>,
        override: Constructable<TOverride>
    ) {
        const instance = this.get(override);
        this.set(parent, instance);
    }
}

File: services/energy-monitor/services/container.go
package services

import (
	"github.com/TWilkin/powerpi/common/services"
	"github.com/TWilkin/powerpi/energy-monitor/services/config"
	energyRetrieverFactory "github.com/TWilkin/powerpi/energy-monitor/services/energyretriever/energyretrieverfactory"
	"github.com/TWilkin/powerpi/energy-monitor/services/meter"
)

type EnergyMonitorContainer interface {
	services.CommonContainer
}

type energyMonitorContainer struct {
	services.CommonContainer
}

func NewEnergyMonitorContainer() EnergyMonitorContainer {
	container := &energyMonitorContainer{
		CommonContainer: services.NewCommonContainer(),
	}

	container.Container().Provide(config.NewConfigService)
	container.Container().Provide(meter.NewMeterManager)
	container.Container().Provide(energyRetrieverFactory.NewEnergyRetrieverFactory)

	// Override the common config service with the shutdown config service
	container.SetConfigService(GetService[config.ConfigService](container))

	return container
}

func GetService[TService any](container EnergyMonitorContainer) TService {
	return services.GetService[TService](container)
}
